### DISCUSSION TITLE
Optimizing GPU AMIs for containerized jobs

### DISCUSSION THREAD

**Discussion by ruffsl** [runs-on/runs-on#123] [Q&A]
When building a custom AMI for GPU enabled runner, as documented here:

- https://runs-on.com/guides/building-custom-ami-with-packer/
- https://runs-on.com/features/gpu-runners/

We forked the `runner-images-for-aws` repo to create a minimal image for containerized jobs needing the nvidia container runtime by using "NVIDIA GPU-Optimized AMI" as the source image:

- https://aws.amazon.com/marketplace/pp/prodview-7ikjtg3um26wq
- https://github.com/botsandus/runner-images-for-aws/compare/4304e20a1af3d88f4a9319e929bdb0971825c3a6...a7b1d66a7cd96051173aeccb28d8c36714942381

Yet, one annoyance when modifying the packer templates was that the DiskSize of the source image (128Gb) seems to dictate the minimal DiskSize of the resulting final AMI as well. Are there any further optimizations that could be taken here to reduce the necessary resources to build and/or run these kind of custom AMIs?

**Comment by crohr** [comment]
Yes the source image dictates the final volume size, unless you mess with the volumes before creating the final AMI, but it's hard to automate. 

Could you do the reverse, i.e. start from the RunsOn full image, and only add what's necessary? Since you only need the nvidia container toolkit, it should be fairly easy to install the required dependencies?

I wrote an article about it some time ago, in case that helps - https://runs-on.com/blog/3-how-to-setup-docker-with-nvidia-gpu-support-on-ubuntu-22/

**Reply by ruffsl** [reply]
I'll give this a shot and report back. I simply started with the `NVIDIA GPU-Optimized AMI` as the source image as I wasn't too familiar with the caveats of AWS's Nvidia enabled EC2 VM instances, and if there was anything unique the official Nvidia AMI's were providing other than an abstraction of the installation of the latest stable kernel drivers and container runtime plugins.

**Reply by crohr** [reply]
@ruffsl would be nice to coordinate with @samayala22, which also opened a similar issue in https://github.com/runs-on/runs-on/discussions/126. Happy to help and provide an official RunsOn base/full image (regularly rebuilt) for GPU usage as well if you end up having a recipe that works for both your use-cases.

**Comment by ruffsl** [comment]
For posterity and future readers, we ended up add these patches to the `runner-images-for-aws` repo:

- https://github.com/botsandus/runner-images-for-aws/pull/1

Most notably these two scripts for installing Nvidia drivers and container toolkit:

```shell
#!/bin/bash -e
################################################################################
##  File:  install-nvidia-driver.sh
##  Desc:  Install nvidia driver onto the image
################################################################################

# Source the helpers for use with the script
source $HELPER_SCRIPTS/install.sh

DEBIAN_FILE="cuda-keyring_1.1-1_all.deb"
REPO_URL="https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2404/x86_64/$DEBIAN_FILE"
GPG_KEY="/usr/share/keyrings/cuda-archive-keyring.gpg"
REPO_PATH="/etc/apt/sources.list.d/cuda-ubuntu2404-x86_64.list"

wget $REPO_URL
sudo dpkg -i $DEBIAN_FILE
apt-get update

# Install nvidia container toolkit which available via apt-get
# Using toolsets keep installation order to install dependencies before the package in order to control versions

get_toolset_value() {
    local toolset_path="${INSTALLER_SCRIPT_FOLDER}/toolset-patch.json"
    local query=$1

    echo "$(jq -r "$query" $toolset_path)"
}

components=$(get_toolset_value '.nvidia_driver.components[] .package')
for package in $components; do
    version=$(get_toolset_value ".nvidia_driver.components[] | select(.package == \"$package\") | .version")
    if [[ $version == "latest" ]]; then
        apt-get install --no-install-recommends "$package"
    else
        version_string=$(apt-cache madison "$package" | awk '{ print $3 }' | grep "$version" | head -1)
        apt-get install --no-install-recommends "${package}=${version_string}"
    fi
done

# Cleanup custom repositories
rm $DEBIAN_FILE
rm $GPG_KEY
rm $REPO_PATH
```

```shell
#!/bin/bash -e
################################################################################
##  File:  install-nvidia-container.sh
##  Desc:  Install nvidia container toolkit onto the image
################################################################################

# Source the helpers for use with the script
source $HELPER_SCRIPTS/install.sh

REPO_URL="https://nvidia.github.io/libnvidia-container/stable/deb/\$(ARCH)"
GPG_KEY="/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg"
REPO_PATH="/etc/apt/sources.list.d/nvidia-container-toolkit.list"

curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | gpg --dearmor -o $GPG_KEY
echo "deb [signed-by=$GPG_KEY] $REPO_URL /" > $REPO_PATH
apt-get update

# Install nvidia container toolkit which available via apt-get
# Using toolsets keep installation order to install dependencies before the package in order to control versions

get_toolset_value() {
    local toolset_path="${INSTALLER_SCRIPT_FOLDER}/toolset-patch.json"
    local query=$1

    echo "$(jq -r "$query" $toolset_path)"
}

components=$(get_toolset_value '.nvidia_container.components[] .package')
for package in $components; do
    version=$(get_toolset_value ".nvidia_container.components[] | select(.package == \"$package\") | .version")
    if [[ $version == "latest" ]]; then
        apt-get install --no-install-recommends "$package"
    else
        version_string=$(apt-cache madison "$package" | awk '{ print $3 }' | grep "$version" | head -1)
        apt-get install --no-install-recommends "${package}=${version_string}"
    fi
done

# Configure the container runtime by using the nvidia-ctk command
nvidia-ctk runtime configure --runtime=docker

# Restart the Docker daemon
systemctl restart docker
docker info

# Cleanup custom repositories
rm $GPG_KEY
rm $REPO_PATH
```

```json
{
  "nvidia_container": {
    "components": [
      {
        "package": "nvidia-container-toolkit",
        "version": "latest"
      }
    ]
  },
  "nvidia_driver": {
    "components": [
      {
        "package": "cuda-drivers",
        "version": "latest"
      }
    ]
  }
}
```

**Comment by ruffsl** [comment]
Update:
- https://github.com/runs-on/runner-images-for-aws/pull/14#event-15857583296

**Reply by crohr** [reply]
Yes it's now live and can be used with standard `ubuntu22-full-x64` or `ubuntu24-full-x64` image names. I'm going to let them live like that for a few days / weeks to see the impact on boot time in the real world, so I might still end up moving the change into a separate image name if boot times degrade too much.

**Reply by crohr** [reply]
This is the boot times for the last 30 days across all users:

![Boot times](https://github.com/user-attachments/assets/2601014e-8c79-4c18-81e9-003edac91afc)

We can clearly see a bump on Jan 9 when I released the new image with the GPU drivers, losing about 3s of boot time.

I think it might actually be better if I release it as a separate image (although more costly to maintain), so that the corresponding packer recipe can also serve as a reference point for users that want a GPU image with more driver versions or specific ML software.

**Reply by ruffsl** [reply]
Sounds reasonable.

As long as the most recent OS nvidia driver and container runtime plugin is included, then that would still be quite convenient with not having to build and host our own AMIs. I've never had an issue with backwards compatibility using cuda drivers, but also having the option to fork and tweak the packer templates with nvidia configs include could save others time.

On that last note, any chance of making the repo's actions more amenable to forking and self hosting via greater global parameters? Had to do a lot of hunting and careful find/replace steps to self-host this before, but will likely encounter merge conflicts with upstream soon given the number of line changes.

- https://github.com/botsandus/runner-images-for-aws/pull/1

Let me know once you have a name for  the new AMIs. I'll just pin things down for now.

**Comment by crohr** [comment]
So the new GPU images have been available for a while (`ubuntu22-gpu-x64` and `ubuntu24-gpu-x64`, and include the cuda drivers + cuda toolkit + nvidia container toolkit.
